Assets/Tests/Runtime/ConnectionManagementTests.cs

```diff
@@ -0,0 +1,593 @@
+            {
+                var expectedClientConnectionStateSequence = new List<ConnectionState>();
+                expectedClientConnectionStateSequence.Add(m_ClientConnectionManagers[i].m_ClientConnecting);
+                expectedClientConnectionStateSequence.Add(m_ClientConnectionManagers[i].m_ClientConnected);
+                expectedClientConnectionStateSequence.Add(m_ClientConnectionManagers[i].m_DisconnectingWithReason);
+                expectedClientConnectionStateSequence.Add(m_ClientConnectionManagers[i].m_Offline);
+                Assert.AreEqual(expectedClientConnectionStateSequence, m_ClientConnectionStateSequences[i]);
+            }
+        }
+
+        [UnityTest]
+        public IEnumerator AttemptingToConnectWithSamePlayerId_ClientsDisconnectedWithReason()
+        {
+            StartHost();
+            Assert.IsTrue(m_ServerNetworkManager.IsHost);
+
+            for (var i = 0; i < NumberOfClients; i++)
+            {
+                var profileManager = m_ClientScopes[i].Container.Resolve<ProfileManager>();
+                profileManager.Profile = $"Client";
+            }
+
+            yield return ConnectClients();
+
+            Assert.IsTrue(m_ClientNetworkManagers[0].IsConnectedClient);
+            for (var i = 1; i < NumberOfClients; i++)
+            {
+                Assert.IsFalse(m_ClientNetworkManagers[i].IsConnectedClient);
+            }
+
+            var expectedServerConnectionStateSequence = new List<ConnectionState>();
+            expectedServerConnectionStateSequence.Add(m_ServerConnectionManager.m_StartingHost);
+            expectedServerConnectionStateSequence.Add(m_ServerConnectionManager.m_Hosting);
+            Assert.AreEqual(expectedServerConnectionStateSequence, m_ServerConnectionStateSequence);
+
+            for (var i = 0; i < NumberOfClients; i++)
+            {
+                var expectedClientConnectionStateSequence = new List<ConnectionState>();
+                expectedClientConnectionStateSequence.Add(m_ClientConnectionManagers[i].m_ClientConnecting);
+                if (i == 0)
+                {
+                    expectedClientConnectionStateSequence.Add(m_ClientConnectionManagers[i].m_ClientConnected);
+                }
+                else
+                {
+                    expectedClientConnectionStateSequence.Add(m_ClientConnectionManagers[i].m_DisconnectingWithReason);
+                    expectedClientConnectionStateSequence.Add(m_ClientConnectionManagers[i].m_Offline);
+                }
+                Assert.AreEqual(expectedClientConnectionStateSequence, m_ClientConnectionStateSequences[i]);
+            }
+        }
+
+        [UnityTest]
+        public IEnumerator ClientConnectingWithoutHost_ConnectionFailed()
+        {
+            SetUniqueProfilesForEachClient();
+
+            for (var i = 0; i < NumberOfClients; i++)
+            {
+                LogAssert.Expect(LogType.Error, k_FailedToConnectToServerErrorMessage);
```

Comment:
you usually don't want to test for string error messages, but for exception types instead. error messages are very likely to change, less for class types.

Reply:
logs are there to help you debug. Why are we testing for logs? If there are no exceptions and just error logs to help you debug, you shouldn't test for this. If it fails later, because we changed our logs, this will just create low priority noise.

Reply:
What I meant is that these error logs already happen during those tests. These make tests fail if they are not expected ([see here](https://docs.unity3d.com/Packages/com.unity.test-framework@1.1/api/UnityEngine.TestTools.LogAssert.html)), which is why we must add this here.